audio/src/main.zig

```diff
@@ -0,0 +1,103 @@
+const std = @import("std");
+const mem = std.mem;
+const testing = std.testing;
+const builtin = @import("builtin");
+const Backend = switch (builtin.os.tag) {
+    .linux,
+    .windows,
+    .macos,
+    .ios,
+    => @import("soundio.zig"),
+    else => @compileError("unsupported os"),
+};
+pub const Error = Backend.Error;
+pub const Device = Backend.Device;
+pub const DeviceIterator = Backend.DeviceIterator;
+
+pub const Mode = enum {
+    input,
+    output,
+};
+
+pub const Format = enum {
+    U8,
+    S16,
+    S24,
+    S32,
+    F32,
+};
+
+pub const DeviceDescriptor = struct {
+    mode: ?Mode = null,
+    format: ?Format = null,
+    is_raw: ?bool = null,
+    channels: ?u8 = null,
+    sample_rate: ?u32 = null,
+    id: ?[]const u8 = null,
+    name: ?[]const u8 = null,
+    callback: ?fn (device: Device, frame_count: u32) void = null,
+};
+
+const Audio = @This();
+
+backend: Backend,
+
+pub fn init(allocator: mem.Allocator) Error!Audio {
+    return Audio{
+        .backend = try Backend.init(allocator),
+    };
+}
+
+pub fn deinit(self: Audio) void {
+    self.backend.deinit();
+}
+
+pub fn waitEvents(self: Audio) void {
+    self.backend.waitEvents();
+}
+
+pub fn requestDevice(self: Audio, config: DeviceDescriptor) Error!Device {
+    return self.backend.requestDevice(config);
+}
+
+pub fn inputDeviceIterator(self: Audio) DeviceIterator {
+    return self.backend.inputDeviceIterator();
+}
+
+pub fn outputDeviceIterator(self: Audio) DeviceIterator {
+    return self.backend.outputDeviceIterator();
+}
+
+test "list devices" {
+    const a = try init(testing.allocator);
+    defer a.deinit();
+
+    var iter = a.inputDeviceIterator();
+    while (try iter.next()) |d| {
+        if (d.id) |id|
+            testing.allocator.free(id);
+    }
+}
+
+test "connect to device" {
+    const a = try init(testing.allocator);
+    defer a.deinit();
+
+    const d = try a.requestDevice(.{ .mode = .output });
+    defer d.deinit();
+}
+
+test "connect to device from id" {
+    const a = try init(testing.allocator);
+    defer a.deinit();
+
+    var iter = a.outputDeviceIterator();
+    var device_desc = (try iter.next()) orelse return error.NoDeviceFound;
+    defer testing.allocator.free(device_desc.id.?);
+
+    const d = try a.requestDevice(device_desc);
```

Comment:
My thinking (let's chat about this in Matrix):

We should create a new descriptor here, passing just `.{.id = device_desc.id, .mode = .output}` to `requestDevice`. Otherwise we're not sure if we are testing it looking up device by ID or by full descriptor info.

Additionally, we should create another test which verifies what happens when the ID doesn't match any known device. My thinking here is that if you pass `.{.id = "does-not-match-any-existing-device-id", .mode = .output}` for a reason like:

* The user literally doesn't have that device anymore (they unplugged it)
* We switched from soundio to a non-soundio backend and the ID format changed
* ...

Then `requestDevice` could see clearly that the device with that ID no longer exists, and instead give you one that you would get if you just passed `requestDevice(.{.mode = .output})` without the ID field.